opt: exploration rule for MIN(x) on index

madhavsuresh opened the pull request:
New MIN(x) index transformation rule replaces MIN(X) with limit 1 in
ascending order. When an (ordered) index is present on the column the
MIN(x) is applied to, a full scan of the underlying table is unecessary.

Along with the rule, additional custom functions were required:

-optgen does not provide declaration of constants, a custom function is
 required to generate an Int Datum in the memo

-The column which the MIN(x) is applied to must be extracted to use for
limit's row ordering

Release note: None

Commit:
opt: exploration rule for MIN(x) on index

New MIN(x) index transformation rule replaces MIN(X) with limit 1 in
ascending order. When an (ordered) index is present on the column the
MIN(x) is applied to, a full scan of the underlying table is unecessary.

Along with the rule, additional custom functions were required:

-optgen does not provide declaration of constants, a custom function is
 required to generate an Int Datum in the memo.

-The column which the MIN(x) is applied to must be extracted to use for
limit's OrderingChoice property

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26905)
<!-- Reviewable:end -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=26905) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=26905) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=26905) it.</sub>

RaduBerinde commented:
Very cool, nice work! Only very minor comments.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFe5m8J9Omgu2MfWA:-LFZFe5m8J9Omgu2MfWB:by8vr4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L331)):*
> ```
> ·          limit        1       ·          ·
> 
> # TODO(radu): we don't yet optimize this by scanning one entry from the xy index.
> ```

Remove this TODO

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFq3t0aIkqe7Lxa7T:-LFZFq3t0aIkqe7Lxa7U:bh7tcio) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
> ```Go
> }
> 
> func (c *CustomFuncs) GetConstOfOne() memo.PrivateID {
> ```

These need a comment (has to start with the function name) or the linter will complain.
Consider renaming to `MakeConstOfOne`, `MakeOrdering`.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFw7H7FJLJ0IrSHoH:-LFZFw7H7FJLJ0IrSHoI:b-jgcbq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L4)):*
> ```
> # groupby.opt contains exploration rules for the groupby operators
> # =============================================================================
> #min on ordered index value
> ```

[nit] Make this into a sentence.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZG5mRBdrVeQQLQPJF:-LFZG5mRBdrVeQQLQPJG:b-shfy2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L5)):*
> ```
> # =============================================================================
> #min on ordered index value
> [MinIndexLimit, Explore]
> ```

Even though an index is the main usecase, there's nothing specific about indexes in this rule, maybe just `MinToLimit`.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZGgh_7t7kiFR9fggb:-LFZGgh_7t7kiFR9fggc:b-4gntux) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
> ```
>  └── aggregations [outer=(1)]
>       └── min [type=int, outer=(1)]
>            └── variable: xyz.x [type=int, outer=(1)]
> ```

[nit] it would be cool to have a cased which shows the `memo` for one of these

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFZGLlnB2kKi4SBMc8n:biv5kmg)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
There are some logictests that failed. You can run them with [lt](https://github.com/RaduBerinde/config-stuff/blob/master/bin/lt), e.g. `lt opt/distsql_agg`

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZHafY4zV53G19PDMS:-LFZHafY4zV53G19PDMT:b-sqwbz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L9)):*
> ```
>   $input:*
>   $aggs: (Aggregations [(Min (Variable $ex:*))] *)
>   $def:*
> ```

We have to check that this is a "scalar" groupby (no grouping columns). This is not a valid transformation if we have `SELECT MIN(x) FROM t GROUP BY y`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFZH8Li3nGAUccWpHWc:bbd9cz6)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Awesome job! I think you need a few more tests, but otherwise looks pretty good.

---

Reviewed 6 of 6 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDFvr0nscH7FDtB2B:-LFZDFvr0nscH7FDtB2C:bcm8fm1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L4)):*
> ```
> # groupby.opt contains exploration rules for the groupby operators
> # =============================================================================
> #min on ordered index value
> ```

[nit] should have a space in front of the comment, and comment should start with `MinIndexLimit`

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDXLvAWc1BOQvyF1y:-LFZDXLvAWc1BOQvyF1z:b4zn0cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L8)):*
> ```
> (GroupBy
>   $input:*
>   $aggs: (Aggregations [(Min (Variable $ex:*))] *)
> ```

I'd rename `ex` to `col` so it's clearer what we are doing here

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZGgYF2c_LMYPER0NH:-LFZGgYF2c_LMYPER0NI:b-x3se7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L13)):*
> ```
> =>
> (GroupBy
>   (Limit $input (Const (GetConstOfOne)) (GetOrdering $ex))
> ```

[nit] I'd suggest constructing the Const inside the custom function. So this would just be `(Limit $input (GetConstOfOne) (GetOrdering $ex))`. Not a big deal either way, but seems a bit cleaner to me.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDrzt4IRTauB7-goE:-LFZDrzt4IRTauB7-goF:b-48l7ke) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
> ```
>  └── aggregations [outer=(1)]
>       └── min [type=int, outer=(1)]
>            └── variable: xyz.x [type=int, outer=(1)]
> ```

It would be good to add some additional test cases, both in which the optimization does or doesn't apply (e.g., there is no index on the variable inside `min`, if there are multiple aggregations, grouping columns, etc)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFZH_fZ7Sj4nk0M7Mnl:bv3xzfd)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Great start to your internship! As Radu and Rebecca said, the content looks good. We set the bar high though, so there's an extra level of polish and finish we always strive for:

- descriptive, consistent naming
- comments on all exported methods and rules, plus on any methods that are tricky in some way
- thorough tests, including positive and negative cases

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZNX2M7LOe1Ao32u04:-LFZNX2M7LOe1Ao32u05:b-jh98ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L336)):*
> ```
> ----
> group      ·            ·                  (min int)  ·
>  │         aggregate 0  min(x)             ·          ·
> ```

Radu, is it worth mapping the `Min` operator to `AnyNotNull`? How come we don't do that in the 2.0 planner?

---

*[pkg/sql/opt/xform/custom_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZSlds27o-cDtPhwrb:-LFZSlds27o-cDtPhwrc:bl0vm14) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L321)):*
> ```Go
> 	return lookupJoinDef.IsIndexJoin(c.e.mem.Metadata())
> }
> ```

Add a section for `GroupBy` rules, using the `Limit Rules` comment above as a template.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZOBkM9jHmtOGr42xT:-LFZOBkN361ZRjLZgw85:bqvjtz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
> ```Go
> }
> 
> func (c *CustomFuncs) GetConstOfOne() memo.PrivateID {
> ```

This is generic enough that it should go into `xfunc/custom_funcs.go`, since we might reuse it in both normalization and exploration rules.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFq3t0aIkqe7Lxa7T:-LFZOsgL7oBTZckYLPMT:b-nq297i) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

These need a comment (has to start with the function name) or the linter will complain.
Consider renaming to `MakeConstOfOne`, `MakeOrdering`.
</blockquote></details>

I'd go with `MakeOne` (we already have `IsOne`), and `MakeOrderingFromColumn`. I think `GetOrdering` and `MakeOrdering` are too generic and therefore easy to confuse with other related methods.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZG5mRBdrVeQQLQPJF:-LFZPLt20pZsw4xCJHrg:be91d45) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L5)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Even though an index is the main usecase, there's nothing specific about indexes in this rule, maybe just `MinToLimit`.
</blockquote></details>

Rules should always start with a verb, so maybe `ReplaceMinWithLimit`.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDXLvAWc1BOQvyF1y:-LFZQBec6UWm5gYkHPxC:bzb2hrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L8)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I'd rename `ex` to `col` so it's clearer what we are doing here
</blockquote></details>

Also, no need for the trailing *, since it's implied. Also, a style nit: get rid of the space after `$aggs`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFZRkrj9aQTqRz4hOiH:blbqbmk)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZNX2M7LOe1Ao32u04:-LFbUgb-AXqShHOfsgat:b-w3okgz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L336)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Radu, is it worth mapping the `Min` operator to `AnyNotNull`? How come we don't do that in the 2.0 planner?
</blockquote></details>

I doubt it matters, the implementations would be doing pretty much the same thing for one row.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Hi Madhav, I just realized that this needs to be a bit more complicated. We don't want the minimum value, we want the minimum not-NULL value (NULL sorts first). We should add an IS NOT NULL SelectOp on top of the input. In addition, we should have a normalization rule which removes any IS NOT NULL filter for columns that are known to be not-null in the logical properties. Also, make sure there are logictests which verify this NULL handling.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFc6VFf-sE-g2olCEOR:b-8q1mu0)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Actually, what @andy-kimball suggests is to have a normalization rule which always adds a not-null filter when there is a a single aggregation that ignores NULLs. Then your rule would apply as-is and things would be simpler (the filter will get merged or pushed down during normalization).

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFc8gV98dWsOJF8uWRn:b-ficsnf)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xfunc/custom_funcs.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFc9fczEvTa07X2BjdH:-LFc9fczEvTa07X2BjdI:b-dnvho2) ([raw file](https://github.com/cockroachdb/cockroach/blob/31a61fc27d41aac349a841d4a7d593501ec60f86/pkg/sql/opt/xfunc/custom_funcs.go#L165)):*
> ```Go
> // EmptyGroupDef returns if the GroupByDef GroupingCols is empty
> // This assumes is that empty GroupingCols implies empty Ordering list
> func (c *CustomFuncs) EmptyGroupByDef(def memo.PrivateID) bool {
> ```

There's an existing function to do this: `IsScalarGroupBy`. You can move that function to `xfunc` so you can share it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/custom_funcs.go, line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFcA-tJ12ur9PzQLmNX:-LFc9aO40LIE6DYV-msW:b-203gks) ([raw file](https://github.com/cockroachdb/cockroach/blob/31a61fc27d41aac349a841d4a7d593501ec60f86/pkg/sql/opt/xform/custom_funcs.go#L338)):*
> ```Go
> // EmptyGroupDef returns if the GroupByDef GroupingCols is empty
> // This assumes is that empty GroupingCols implies empty Ordering list
> func (c *CustomFuncs) EmptyGroupByDef(def memo.PrivateID) bool {
> ```

Duplicated function.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Though even if we add this other normalization rule, I'd still add a condition to the MIN rule so that it only fires when the aggregate column is not-null. Otherwise, the rule could mistakenly fire if we have optimizations disabled.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFcFa_g2O9Bvm9TJQZy:b8er6u0)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Working on a supporting PR for removing `NOT NULL` filter predicate applied to columns where `NOT NULL` is an existing logical property of the column. 

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFrMHWHDY98Tj_-chpb:b-pcx4oq)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFe5m8J9Omgu2MfWA:-LFy5sjMAAGXBT560Xgw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L331)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Remove this TODO
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 321 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZSlds27o-cDtPhwrb:-LFy62o17p2PdFcxkSd0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L321)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add a section for `GroupBy` rules, using the `Limit Rules` comment above as a template.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFq3t0aIkqe7Lxa7T:-LFcDSZy4aG5agesOORv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'd go with `MakeOne` (we already have `IsOne`), and `MakeOrderingFromColumn`. I think `GetOrdering` and `MakeOrdering` are too generic and therefore easy to confuse with other related methods.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZOBkM9jHmtOGr42xT:-LFcDQze2yDr05zPTVqZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This is generic enough that it should go into `xfunc/custom_funcs.go`, since we might reuse it in both normalization and exploration rules.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 338 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFcA-tJ12ur9PzQLmNX:-LFcDZ3s6VOcTBzOPMYC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31a61fc27d41aac349a841d4a7d593501ec60f86/pkg/sql/opt/xform/custom_funcs.go#L338)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Duplicated function.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDFvr0nscH7FDtB2B:-LFcE6VV8gB-0CY_K63h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L4)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] should have a space in front of the comment, and comment should start with `MinIndexLimit`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZFw7H7FJLJ0IrSHoH:-LFcE5iP-Qo9zRSYSiky:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L4)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Make this into a sentence.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZG5mRBdrVeQQLQPJF:-LFcE71v2NIc6bndiPDe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L5)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Rules should always start with a verb, so maybe `ReplaceMinWithLimit`.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 8 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDXLvAWc1BOQvyF1y:-LFcE8TMAmLN-bz5mIWU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L8)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Also, no need for the trailing *, since it's implied. Also, a style nit: get rid of the space after `$aggs`.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZHafY4zV53G19PDMS:-LFy6724BeZhzFAEY2iI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L9)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We have to check that this is a "scalar" groupby (no grouping columns). This is not a valid transformation if we have `SELECT MIN(x) FROM t GROUP BY y`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZGgYF2c_LMYPER0NH:-LFy67iLDallNbpbPkgw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/rules/groupby.opt#L13)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] I'd suggest constructing the Const inside the custom function. So this would just be `(Limit $input (GetConstOfOne) (GetOrdering $ex))`. Not a big deal either way, but seems a bit cleaner to me.
</blockquote></details>

Done.

---

*[pkg/sql/opt/xfunc/custom_funcs.go, line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFc9fczEvTa07X2BjdH:-LFy6GdR81dTAHzKsYhj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/31a61fc27d41aac349a841d4a7d593501ec60f86/pkg/sql/opt/xfunc/custom_funcs.go#L165)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

There's an existing function to do this: `IsScalarGroupBy`. You can move that function to `xfunc` so you can share it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Need to extend test coverage. In particular, logic tests need to be added.

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LFy6sv5DAnYXphmJjyO:b-r52esz)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 1 of 5 files at r2, 6 of 6 files at r3.
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZOBkM9jHmtOGr42xT:-LG0TwTHDm_PEEAolmhY:bu237bs) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks like that change didn't get pushed

---

*[pkg/sql/opt/xform/custom_funcs.go, line 466 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0TSYT8ngaM55zMsma:-LG0TSYT8ngaM55zMsmb:b-1zvrp4) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/custom_funcs.go#L466)):*
> ```Go
> // ----------------------------------------------------------------------
> 
> // MakeConstOfOne returns a memo.PrivateID pointing
> ```

MakeConstOfOne -> MakeOne

Also returns a memo.GroupID, not a memo.PrivateID anymore

Finish comment with a period

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG0UmMV4pRG6QfxEkxB:bjdzj4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
> ```
>   (Limit
>     (Select
>     $input
> ```

This should be indented

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG0UzwX2UgZeTtRN7f6:b87fx2a) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
> ```
>     (Select
>     $input
>     (Filters [(IsNot (Variable $col) (Null (AnyType)))])
> ```

If you bind `(Variable $col:*)` in the Match part of the rule to a variable, you don't need to reconstruct it here

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 19 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0VMzT-bYYoBJdyG-2:-LG0VMzT-bYYoBJdyG-3:b-np63p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L19)):*
> ```
>     (Filters [(IsNot (Variable $col) (Null (AnyType)))])
>     )
>   (MakeOne) (MakeOrderingFromColumn $col)
> ```

should be indented. I'd also put `(MakeOne)` and `(MakeOrderingFromColumn $col)` on separate lines

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 21 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0VXEI5o_VKxJPxFca:-LG0VXEI5o_VKxJPxFcb:b-et37on) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L21)):*
> ```
>   (MakeOne) (MakeOrderingFromColumn $col)
>   )
>   (Aggregations [(AnyNotNull (Variable $col))] $cols)
> ```

And again here - reuse a variable bound from the Match part of the rule for `(Variable $col)`

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZOBkM9jHmtOGr42xT:-LFZgrMTAGUiokimJtVB:b2m6of) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Looks like that change didn't get pushed
</blockquote></details>

Done was the wrong comment.  xfunc/custom_funcs.go doesn't have access to norm.Factory so this function has to stay within either norm/custom_funcs.go or xform/custom_funcs.go

---

*[pkg/sql/opt/xform/custom_funcs.go, line 466 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0TSYT8ngaM55zMsma:-LG0XwDS-RmCUzUvV5UI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/custom_funcs.go#L466)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

MakeConstOfOne -> MakeOne

Also returns a memo.GroupID, not a memo.PrivateID anymore

Finish comment with a period
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG0Yc3Q04F9lb_SovlW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This should be indented
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG0YcSI2HjAcVdCgkvZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

If you bind `(Variable $col:*)` in the Match part of the rule to a variable, you don't need to reconstruct it here
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 19 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0VMzT-bYYoBJdyG-2:-LG0Yd1SAWBeQpfKqgo-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L19)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

should be indented. I'd also put `(MakeOne)` and `(MakeOrderingFromColumn $col)` on separate lines
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 21 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0VXEI5o_VKxJPxFca:-LG0YdnmBDmm2z6yltAH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L21)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

And again here - reuse a variable bound from the Match part of the rule for `(Variable $col)`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZGgh_7t7kiFR9fggb:-LG0_dOV6oS-4w1ymEbP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] it would be cool to have a cased which shows the `memo` for one of these
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDrzt4IRTauB7-goE:-LG0fkzs5CiqZ-cgFJhS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It would be good to add some additional test cases, both in which the optimization does or doesn't apply (e.g., there is no index on the variable inside `min`, if there are multiple aggregations, grouping columns, etc)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

---

Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/custom_funcs.go, line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0mPwH3TaDcpcxuez3:-LG0mPwH3TaDcpcxuez4:b-5apxyj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/custom_funcs.go#L24)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
> 	"github.com/cockroachdb/cockroach/pkg/util"
> 	"github.com/cockroachdb/cockroach/pkg/sql/sem/types"
> ```

Hm, are you not using `goimports` or `crlfmt` (https://github.com/cockroachdb/crlfmt)? The linter is complaining about these having to be sorted

---

*[pkg/sql/opt/xform/custom_funcs.go, line 473 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0mLnP0cab3ndVwNYu:-LG0mLnP0cab3ndVwNYv:b-7zqfjs) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/custom_funcs.go#L473)):*
> ```Go
> }
> 
> // MakeOrderingFromColID creates an Ordering for a column based on the an input ColID
> ```

[nit] MakeOrderingFromColumn

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG0m4-Z5rIAT_tVwiif:bcmw7ss)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Review status: :shipit: complete! 1 of 0 LGTMs obtained

---

*[pkg/sql/opt/xform/custom_funcs.go, line 24 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0mPwH3TaDcpcxuez3:-LG0r_YQAKW4dJfGjdxD:bg70a1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/custom_funcs.go#L24)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, are you not using `goimports` or `crlfmt` (https://github.com/cockroachdb/crlfmt)? The linter is complaining about these having to be sorted
</blockquote></details>

Yikes, I am not. I made a faulty assumption that GoLand was auto-formatting on save. Will figure that out. 

---

*[pkg/sql/opt/xform/custom_funcs.go, line 473 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0mLnP0cab3ndVwNYu:-LG0s5UQ5-FWIEzv7rXT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/custom_funcs.go#L473)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] MakeOrderingFromColumn
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 4 of 4 files at r4, 1 of 1 files at r5.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/custom_funcs.go, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZOBkM9jHmtOGr42xT:-LG0sQhe2wjIKJ4xDD9i:b-sm6sj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/custom_funcs.go#L322)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Done was the wrong comment.  xfunc/custom_funcs.go doesn't have access to norm.Factory so this function has to stay within either norm/custom_funcs.go or xform/custom_funcs.go
</blockquote></details>

Good point - thanks for clarifying!

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG0s_t-7u0ivLANHPDw:bbjuqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think most of the other rules are indenting 4 spaces rather than two (just make sure this is consistent with whatever the other rules are doing)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG0syUC86Hq9namOGwI:bspq9fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks good. I'd just name them "var" and "col" for simplicity.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDrzt4IRTauB7-goE:-LG0v4R8DctW0glco7Jy:b-drmxq0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Nice tests! Can you add one more test like this:

`SELECT min(y) FROM xyz GROUP BY x`

Just to ensure that even though min is by itself, the presence of a grouping column prevents the optimization.

And something like this:
`SELECT min(y), min(x) FROM xyz` which doesn't work since the min is applied to different columns

Maybe also something like this:
`SELECT min(y), min(y) FROM xyz` Here the optimization *should* apply, because the optbuilder will consolidate the duplicate aggregations

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0vQWK7T15yFg9b_RD:-LG0vQWK7T15yFg9b_RE:b3zmh3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/testdata/rules/groupby#L131)):*
> ```
>            └── variable: xyz.y [type=int, outer=(2)]
> 
> # ReplaceMinWithLimit does not apply
> ```

[nit] I'd mention that it doesn't apply because there are group by columns

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0v_ps53_syUp45c72:-LG0v_ps53_syUp45c73:b-ifgt8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/testdata/rules/groupby#L152)):*
> ```
>                 └── variable: xyz.y [type=int, outer=(2)]
> 
> # ReplaceMinWithLimit does not apply
> ```

[nit] mention reason it doesn't apply

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG10MIH6f4c0kYAlvBE:b-9mrmax) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think most of the other rules are indenting 4 spaces rather than two (just make sure this is consistent with whatever the other rules are doing)
</blockquote></details>

It seems to be a little bit of both. For example, select.opt uses all two space indenting, limit.opt uses four spaces sometimes, two spaces other times (sometimes in the same rule). Is there a style guide for the *.opt rules

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG0xeFt46mPgizlhJtc:b-ey0368) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Looks good. I'd just name them "var" and "col" for simplicity.
</blockquote></details>

The compiler is doesn't like using the variable $var (var is reserved in go?) 

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 93 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LFZDrzt4IRTauB7-goE:-LG1-vaE-fOgxJYAVODV:brritei) ([raw file](https://github.com/cockroachdb/cockroach/blob/6200aa2963017c0c5440020ce477077c231175f1/pkg/sql/opt/xform/testdata/rules/groupby#L93)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Nice tests! Can you add one more test like this:

`SELECT min(y) FROM xyz GROUP BY x`

Just to ensure that even though min is by itself, the presence of a grouping column prevents the optimization.

And something like this:
`SELECT min(y), min(x) FROM xyz` which doesn't work since the min is applied to different columns

Maybe also something like this:
`SELECT min(y), min(y) FROM xyz` Here the optimization *should* apply, because the optbuilder will consolidate the duplicate aggregations
</blockquote></details>

Thanks for the suggestions, added them!

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 131 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0vQWK7T15yFg9b_RD:-LG1-tUA-h5UcSfXtgNX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/testdata/rules/groupby#L131)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] I'd mention that it doesn't apply because there are group by columns
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/testdata/rules/groupby, line 152 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0v_ps53_syUp45c72:-LG1-uhR-w48yazsJ-wS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae660d33cf5a7593d383678c50fd0d557e634120/pkg/sql/opt/xform/testdata/rules/groupby#L152)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] mention reason it doesn't apply
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG1JqqCBui7M49vd0Wv:b-p0xhdg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

It seems to be a little bit of both. For example, select.opt uses all two space indenting, limit.opt uses four spaces sometimes, two spaces other times (sometimes in the same rule). Is there a style guide for the *.opt rules
</blockquote></details>

@andy-kimball is our style guide :)  I also thought it was 4 spaces (though maybe we should make it 2 since that's what we use for go code). We should at least be consistent within a rule though (the "match" side is at 4 and the "replace" side is at 2 now).

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG1LTQO3epskyzTADpe:bx7ubcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

@andy-kimball is our style guide :)  I also thought it was 4 spaces (though maybe we should make it 2 since that's what we use for go code). We should at least be consistent within a rule though (the "match" side is at 4 and the "replace" side is at 2 now).
</blockquote></details>

Got it. I've updated the rule to have 4 space indents. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UmMV4pRG6QfxEkxA:-LG1Ot7R4pNiEDeOdynf:b-z4liy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L16)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

Got it. I've updated the rule to have 4 space indents. 
</blockquote></details>

We should probably switch to use tabs. I always set my editor so that tabs=4 spaces, b/c I find it harder to read code using 2 spaces. But if we used tabs, then everyone can set their editor at either 2 or 4 spaces according to their preference.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Reviewed 2 of 2 files at r6.
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG1Pkyh6COXhUZomxV4:bibw9u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
<details><summary><i>Previously, madhavsuresh wrote…</i></summary><blockquote>

The compiler is doesn't like using the variable $var (var is reserved in go?) 
</blockquote></details>

What about just $variable? (I'm also fine with $minvar - up to you)

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG1QsQ37mLVSS58Rd_2:b-uwkowd)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
Updated after rebase - OrderingChoice is now used for Limit. 

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/rules/groupby.opt, line 17 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG0UzwX2UgZeTtRN7f5:-LG1WXek0EJ2N0U2RmoD:b-u6rawm) ([raw file](https://github.com/cockroachdb/cockroach/blob/bdd1e3ed4ecf6f551ada266131b0ecc9e73bbe2c/pkg/sql/opt/xform/rules/groupby.opt#L17)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

What about just $variable? (I'm also fine with $minvar - up to you)
</blockquote></details>

Updated it - $variable is more consistent with the naming scheme of current rules. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG1zCxXC2fnR6ig6ZxR:bqnkzah)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
bors r+

---

Reviewed 1 of 4 files at r4, 1 of 5 files at r7.
Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG24RwZ7qu9zxctsyDW:b-ydi9ay)*
<!-- Sent from Reviewable.io -->

rytaft commented:
*[pkg/sql/opt/xform/custom_funcs.go, line 464 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG26YxjEBGEZ4cuR8pz:-LG26YxkDlrJCHs_PgvZ:b1qqiqe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f38fb5c9e353b72baaadbe03c2747d70af4e57d/pkg/sql/opt/xform/custom_funcs.go#L464)):*
> ```Go
> func (c *CustomFuncs) MakeAscOrderingChoiceFromColumn(col memo.PrivateID) memo.PrivateID {
> 	oc := props.OrderingChoice{}
> 	oc.AppendCol(c.ExtractColID(col), false)
> ```

when you pass a bare value like this, the style guide is to add a comment like `false /* descending */` 

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
bors r+

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---

*[pkg/sql/opt/xform/custom_funcs.go, line 461 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG27HwtAcXJxfzexm8I:-LG2aO_4-x07mmh7Rjqg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f38fb5c9e353b72baaadbe03c2747d70af4e57d/pkg/sql/opt/xform/custom_funcs.go#L461)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

[nit] the input ID -> the input ColumnID
[nit] add period to comment
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/custom_funcs.go, line 464 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-LG26YxjEBGEZ4cuR8pz:-LG2aV_yCcoLTezqr6q4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f38fb5c9e353b72baaadbe03c2747d70af4e57d/pkg/sql/opt/xform/custom_funcs.go#L464)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

when you pass a bare value like this, the style guide is to add a comment like `false /* descending */` 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG2a_dZ4G1XoToUX0Rx:b-ydi9ay)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
bors r-

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG2ad7m8EglgH-vF9u7:b-ydi9aw)*
<!-- Sent from Reviewable.io -->

madhavsuresh commented:
bors r+

---

Review status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26905#-:-LG2qP6h3mmj0txRs6Bc:b-ydi9ay)*
<!-- Sent from Reviewable.io -->
